src/Appwrite/Migration/Version/V19.php

```diff
@@ -0,0 +1,134 @@
+                $name,
+                fn () => null,
+                fn () => []
+            );
+        }
+
+        Console::log('Migrating Project: ' . $this->project->getAttribute('name') . ' (' . $this->project->getId() . ')');
+        $this->projectDB->setNamespace("_{$this->project->getInternalId()}");
+
+        $this->alterPermissionIndex('_metadata');
+
+        Console::info('Migrating Databases');
+        $this->migrateDatabases();
+
+        Console::info('Migrating Collections');
+        $this->migrateCollections();
+
+        Console::info('Migrating Buckets');
+        $this->migrateBuckets();
+
+        Console::info('Migrating Documents');
+        $this->forEachDocument([$this, 'fixDocument']);
+    }
+
+    /**
+     * Migrate all Databases.
+     *
+     * @return void
+     * @throws \Exception
+     */
+    private function migrateDatabases(): void
+    {
+        foreach ($this->documentsIterator('databases') as $database) {
+            Console::log("Migrating Collections of {$database->getId()} ({$database->getAttribute('name')})");
+
+            $databaseTable = "database_{$database->getInternalId()}";
+
+            $this->alterPermissionIndex($databaseTable);
+
+            foreach ($this->documentsIterator($databaseTable) as $collection) {
+                $collectionTable = "{$databaseTable}_collection_{$collection->getInternalId()}";
+                Console::log("Migrating Collections of {$collectionTable} {$collection->getId()} ({$collection->getAttribute('name')})");
+                $this->alterPermissionIndex($collectionTable);
+            }
+        }
+    }
+
+    /**
+     * Migrate all Collections.
+     *
+     * @return void
+     */
+    private function migrateCollections(): void
+    {
+        foreach ($this->collections as $collection) {
+            $id = $collection['$id'];
+
+            Console::log("Migrating Collection \"{$id}\"");
+
+            if (!in_array($id, ['files', 'collections'])) {
+                $this->alterPermissionIndex($id);
+            }
+
+            usleep(50000);
+        }
+    }
+
+    /**
+     * Fix run on each document
+     *
+     * @param Document $document
+     * @return Document
+     */
+    protected function fixDocument(Document $document): Document
+    {
+        switch ($document->getCollection()) {
+            case 'projects':
+                /**
+                 * Bump version number.
+                 */
+                $document->setAttribute('version', '1.4.0');
+                break;
+        }
+
+        return $document;
+    }
+
+    protected function alterPermissionIndex($collectionName): void
+    {
+        try {
+            $this->pdo->prepare("ALTER TABLE `{$this->projectDB->getDefaultDatabase()}`.`_{$this->project->getInternalId()}_{$collectionName}_perms` 
+            DROP INDEX `_permission`, 
+            ADD INDEX `_permission` (`_permission`, `_type`, `_document`)
+            ")->execute();
```

Comment:
This will be a bit easier to read:

```suggestion
            $table = "`{$this->projectDB->getDefaultDatabase()}`.`_{$this->project->getInternalId()}_{$collectionName}_perms";
            $this->pdo->prepare("
                ALTER TABLE {$table}
                DROP INDEX `_permission`, 
                ADD INDEX `_permission` (`_permission`, `_type`, `_document`);
            ")->execute();
```